File: gilded-rose/python/src/gilded_rose/gilded_rose_inn.py
from .category import Category
from .inventory import Inventory


class GildedRoseInn:
    def __init__(self, inventory: Inventory) -> None:
        self.inventory = inventory

    def update_inventory(self) -> None:
        for item in self.inventory.items:
            if item.category == Category.LEGENDARY:
                continue
            if item.category == Category.AGED:
                gain = 2 if item.sell_in <= 0 else 1
                item.quality = min(50, item.quality + gain)
            elif item.category == Category.BACKSTAGE_PASS:
                if item.sell_in <= 0:
                    item.quality = 0
                else:
                    if item.sell_in <= 5:
                        gain = 3
                    elif item.sell_in <= 10:
                        gain = 2
                    else:
                        gain = 1
                    item.quality = min(50, item.quality + gain)
            elif item.category == Category.CONJURED:
                degrade = 4 if item.sell_in <= 0 else 2
                item.quality = max(0, item.quality - degrade)
            else:
                degrade = 2 if item.sell_in <= 0 else 1
                item.quality = max(0, item.quality - degrade)
            item.sell_in -= 1

File: gilded-rose/csharp/src/GildedRose/GildedRoseInn.cs
namespace GildedRose;

public class GildedRoseInn
{
    public Inventory Inventory { get; }

    public GildedRoseInn(Inventory inventory)
    {
        Inventory = inventory;
    }

    public void UpdateInventory()
    {
        foreach (var item in Inventory.Items)
        {
            if (item.Category == Category.Legendary)
            {
                continue;
            }
            if (item.Category == Category.Aged)
            {
                var gain = item.SellIn <= 0 ? 2 : 1;
                item.Quality = Math.Min(50, item.Quality + gain);
            }
            else if (item.Category == Category.BackstagePass)
            {
                if (item.SellIn <= 0)
                {
                    item.Quality = 0;
                }
                else
                {
                    var gain = item.SellIn <= 5 ? 3
                        : item.SellIn <= 10 ? 2
                        : 1;
                    item.Quality = Math.Min(50, item.Quality + gain);
                }
            }
            else if (item.Category == Category.Conjured)
            {
                var degrade = item.SellIn <= 0 ? 4 : 2;
                item.Quality = Math.Max(0, item.Quality - degrade);
            }
            else
            {
                var degrade = item.SellIn <= 0 ? 2 : 1;
                item.Quality = Math.Max(0, item.Quality - degrade);
            }
            item.SellIn -= 1;
        }
    }
}
